diff --git a/lib/Makefile b/lib/Makefile
index a4cf61ab..6781277f 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -45,6 +45,13 @@ ifneq ($(ZSTD_LIB_DICTBUILDER), 0)
   ZSTD_FILES += $(ZSTD_DICTBUILDER_FILES)
 endif
 
+QAT_SUPPORT ?= 0
+ifneq ($(QAT_SUPPORT), 0)
+  QAT_INCLUDE = -I$(ICP_ROOT)/quickassist/include -I$(ICP_ROOT)/quickassist/include/dc -I$(ICP_ROOT)/quickassist/lookaside/access_layer/include
+  LDFLAGS += -L$(ICP_ROOT)/build -lqat_s -static -lqatmf
+  CPPFLAGS += -DQAT_SUPPORT $(QAT_INCLUDE)
+endif
+
 ZSTD_LOCAL_SRC := $(notdir $(ZSTD_FILES))
 ZSTD_LOCAL_OBJ0 := $(ZSTD_LOCAL_SRC:.c=.o)
 ZSTD_LOCAL_OBJ := $(ZSTD_LOCAL_OBJ0:.S=.o)
diff --git a/lib/compress/zstd_compress.c b/lib/compress/zstd_compress.c
index b55f684c..aeebcb9b 100644
--- a/lib/compress/zstd_compress.c
+++ b/lib/compress/zstd_compress.c
@@ -27,6 +27,9 @@
 #include "zstd_ldm.h"
 #include "zstd_compress_superblock.h"
 #include  "../common/bits.h"      /* ZSTD_highbit32 */
+#ifdef QAT_SUPPORT
+#include "qatmatchfinder.h"
+#endif
 
 /* ***************************************************************
 *  Tuning parameters
@@ -107,6 +110,9 @@ static void ZSTD_initCCtx(ZSTD_CCtx* cctx, ZSTD_customMem memManager)
         assert(!ZSTD_isError(err));
         (void)err;
     }
+#ifdef QAT_SUPPORT
+    ZSTD_QAT_startQatDevice(cctx);
+#endif
 }
 
 ZSTD_CCtx* ZSTD_createCCtx_advanced(ZSTD_customMem customMem)
@@ -173,6 +179,9 @@ static void ZSTD_freeCCtxContent(ZSTD_CCtx* cctx)
     ZSTDMT_freeCCtx(cctx->mtctx); cctx->mtctx = NULL;
 #endif
     ZSTD_cwksp_free(&cctx->workspace, cctx->customMem);
+#ifdef QAT_SUPPORT
+    ZSTD_QAT_stopQatDevice();
+#endif
 }
 
 size_t ZSTD_freeCCtx(ZSTD_CCtx* cctx)
